_posts/2020-10-27-a-closer-look-at-the-interpreter.md

```diff
@@ -0,0 +1,202 @@
+its arguments, which are in turn followed by the next instruction:
+
+    Instruction address
+        First argument
+        Second argument
+        ... and so on.
+    Instruction address
+        First argument
+        Second argument
+        ... and so on.
+
+When an instruction finishes it reads the next instruction address and jumps
+to it, forever following the "thread" of instructions.
+
+With very few exceptions, the instructions are "pure" in the sense that they
+always do the same thing with the same input and that they only affect BEAM,
+either through changing the control flow or writing a result to a register.
+This makes the instructions very easy to read and reason about in isolation, so
+for the sake of brevity we we'll only have a look at one of them:
+
+```c++
+is_nonempty_list(Label, Src) {
+
+    /* Check if our $Src is not a list. */
+    if (is_not_list($Src)) {
+
+        /* Invoke the $FAIL macro, jumping to our
+         * $Label. */
+        $FAIL($Label);
+    }
+
+    /* Execute the next instruction. */
+}
+```
+
+The above is written in a domain-specific language (DSL) that is a superset of
+C, where `$`-prefixed macros are expanded and the rest is kept as is.
+
+The [`beam_makeops`] script takes this definition and generates code for
+the parts that are very cumbersome to write by hand, such as jumping to the
+next instruction. It also hides argument handling which allows us to reduce
+the code size by packing small arguments together behind the scenes, which
+we've explored in an [earlier post] on the subject.
+
+For performance reasons it also generates different variants based on the
+argument types outlined in [`ops.tab`] to reduce the amount of work we need to
+do at runtime.
+
+Let's have a look at the generated code for this instruction, specifically the
+variant for `X` registers:
+
+```c++
+/* (This has been modified slightly for readability) */
+case is_nonempty_list_fx:
+{
+    Eterm arg_word, term;
+
+    /* Read the argument word from the instruction
+     * stream. */
+    arg_word = I[1];
+
+    /* Unpack the offset of our source register (upper
+     * 32 bits) and then read its contents.
+     *
+     * Note that we address the X registers directly;
+     * had this instruction not been specialized, we
+     * would first need to determine whether the
+     * argument was an X or a Y register. */
+    term = x_registers[arg_word >> 32];
+
+    /* is_not_list(term) */
+    if (term & (_TAG_PRIMARY_MASK - TAG_PRIMARY_LIST)) {
+
+        /* Unpack our fail label (lower 32 bits) and add
+         * it to our current instruction pointer. This
+         * is an offset and may be negative for backward
+         * jumps. */
+        I += (Sint32)arg_word;
+
+        /* Jump to the instruction at the fail label using
+         * the "labels as values" GCC extension. */
+        goto *I;
+    }
+
+    /* Skip the current label address and argument word,
```

Comment:
I think you should mention that this part was automatically generated by `beam_makeops`.

Perhaps you also make the wording similar to the comment in the same place in source instrs file. ("Execute the next instruction.")